### ISSUE TITLE
Student's view:grade items can not be sorted alphabetically

### ISSUE THREAD

**Issue by raulsv** [sakaiproject/sakai#2284]
Enhacement: Grade items cannot be sorted alphabetically in the student's view as you can see in the screenshot attached to this issue.

![order](https://cloud.githubusercontent.com/assets/10761548/14673761/1b03776c-0700-11e6-8947-1df7dc3afb96.png)

**Comment by payten** [comment]
Thanks @raulsv, currently the student listing doesn't seem to sort by anything particularly meaningful.  Unlike the instructor view, which is far cleverer!

@kyleblythe, I'm just thinking through a few scenarios and I think we'll need to clarify a few things before we jump in.

In the instructor view when categories are disabled, we use the sort order defined via the drag drop sorting and stored in the database `gb_gradable_object_t.sort_order`. So the ordering is generally defined by the instructor.  Perhaps we should be honoring this sort order in the student view also?

To nail it down, are we looking to simply default to an alphabetical sort order for this student view? Or offer students the ability to sort the grade table and change between the default ordering and alphabetical ordering?  It feels like a chronological order would be just as useful, perhaps based on 'Due date' or the time the grade was released.

The scenario above covers the 'No Categories' option.  If categories are enabled, should we then alpha-sort the categories and then the items underneath?

**Comment by steveswinsburg** [comment]
@payten @kyleblythe I'm thinking we just reuse whatever sort order the instructor has setup, so that they can arrange things how they want and know it will be the same for students. In fact I thought that was already the case but maybe I wasn't paying close attention :fearful:

**Comment by steveswinsburg** [comment]
I looked at this last night and it should be pretty trivial to use the same sort order as what the instructor has setup, which ideally is a global setting for the entire course (ie they have set it up to how it should be). The questions is whether or not it makes sense for a student to be able to sort it how they want to see it.

**Comment by raulsv** [comment]
The main purpose of this feature request is to allow the student to sort alphabetically automatically because the gradebook structure could be:

Grade 4
Grade 1
Grade 3
Grade 2

**Comment by jeffpasch** [comment]
Thanks, @raulsv. 

We discussed in GradebookNG call. We'll aim to implement sorting for the student as it currently exists in Gradebook 1.

**Comment by steveswinsburg** [comment]
This has been implemented in #2284. Nice one Payten.
I haven't yet checked the printing works with the new sort, @payten?

**Comment by kyleblythe** [comment]
Fantastic - looks great! AND the student has the ability to group by category or not (bonus! Very nice touch!)

![image](https://cloud.githubusercontent.com/assets/11447474/15372002/418de224-1d0c-11e6-9d04-ad5118b2326b.png)

This does not seem to get reflected in the print view though.

**Comment by raulsv** [comment]
Great job!

**Comment by jeffpasch** [comment]
A few small changes:

When not grouped by category, don't show  the Expand All, Collapse All links.
http://www.screencast.com/t/i3a1eHX3

By default, sort by Due Date ascending

**Comment by payten** [comment]
Also, in Chrome, print doesn't capture the state of the panel, so if Categories are not grouped, the print is currently showing the items grouped.

Also add ability to group on the Grade Summary tab and also sync up the Group By Category state with the state of the Gradebook settings value.

**Comment by steveswinsburg** [comment]
Was this one still outstanding or is it done @payten? I can't see another commit referencing it after the last comment.

**Comment by payten** [comment]
Ah oops. I think the only thing remaining was the change to include the "Group by Category" toggle on the Grade Summary tab (the instructor view).  Previously I'd only added this to the student view.

All the other bug fixes have been rolled in via different issues.  I wonder if we can remove the priority tag now that it's behaving well and I can look into enhancing the instructor view once the other priority issues are completed? Does that seem ok @jeffpasch, @kyleblythe?

**Comment by kyleblythe** [comment]
Hmm. Still getting a runtime exception when I try to access the grade summary lightbox. Might be the same issue as was reported in #2732 . Regardless, here's the error I get today:

`org.apache.wicket.WicketRuntimeException: Error attaching this container for rendering: [StudentGradeSummaryPanel [Component id = content]]
	at org.apache.wicket.MarkupContainer.onBeforeRenderChildren(MarkupContainer.java:1696)
	at org.apache.wicket.Component.onBeforeRender(Component.java:3826)
	at org.apache.wicket.Component.internalBeforeRender(Component.java:949)
	at org.apache.wicket.Component.beforeRender(Component.java:1017)
	at org.apache.wicket.MarkupContainer.onBeforeRenderChildren(MarkupContainer.java:1684)
	at org.apache.wicket.Component.onBeforeRender(Component.java:3826)
	at org.apache.wicket.extensions.ajax.markup.html.modal.ModalWindow.onBeforeRender(ModalWindow.java:856)
	at org.apache.wicket.Component.internalBeforeRender(Component.java:949)
	at org.apache.wicket.Component.beforeRender(Component.java:1017)
	at org.apache.wicket.Component.internalPrepareForRender(Component.java:2194)
	at org.apache.wicket.Component.prepareForRender(Component.java:2233)
	at org.apache.wicket.ajax.XmlAjaxResponse.writeComponent(XmlAjaxResponse.java:107)
	at org.apache.wicket.ajax.AbstractAjaxResponse.writeComponents(AbstractAjaxResponse.java:218)
	at org.apache.wicket.ajax.AbstractAjaxResponse.writeTo(AbstractAjaxResponse.java:150)
	at org.apache.wicket.ajax.AjaxRequestHandler.respond(AjaxRequestHandler.java:359)
	at org.apache.wicket.request.cycle.RequestCycle$HandlerExecutor.respond(RequestCycle.java:862)
	at org.apache.wicket.request.RequestHandlerStack.execute(RequestHandlerStack.java:64)
	at org.apache.wicket.request.RequestHandlerStack.execute(RequestHandlerStack.java:97)
	at org.apache.wicket.request.cycle.RequestCycle.execute(RequestCycle.java:261)
	at org.apache.wicket.request.cycle.RequestCycle.processRequest(RequestCycle.java:218)
	at org.apache.wicket.request.cycle.RequestCycle.processRequestAndDetach(RequestCycle.java:289)
	at org.apache.wicket.protocol.http.WicketFilter.processRequestCycle(WicketFilter.java:259)
	at org.apache.wicket.protocol.http.WicketFilter.processRequest(WicketFilter.java:201)
	at org.apache.wicket.protocol.http.WicketServlet.doGet(WicketServlet.java:137)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:622)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:291)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.sakaiproject.util.RequestFilter.doFilter(RequestFilter.java:418)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.catalina.core.ApplicationDispatcher.invoke(ApplicationDispatcher.java:721)
	at org.apache.catalina.core.ApplicationDispatcher.processRequest(ApplicationDispatcher.java:466)
	at org.apache.catalina.core.ApplicationDispatcher.doForward(ApplicationDispatcher.java:358)
	at org.apache.catalina.core.ApplicationDispatcher.forward(ApplicationDispatcher.java:318)
	at org.sakaiproject.tool.impl.ActiveToolComponent$MyActiveTool.forward(ActiveToolComponent.java:513)
	at org.sakaiproject.portal.charon.SkinnableCharonPortal.forwardTool(SkinnableCharonPortal.java:1474)
	at org.sakaiproject.portal.charon.handlers.SiteHandler.doSite(SiteHandler.java:463)
	at org.sakaiproject.portal.charon.handlers.SiteHandler.doGet(SiteHandler.java:263)
	at org.sakaiproject.portal.charon.handlers.SiteHandler.doGet(SiteHandler.java:151)
	at org.sakaiproject.portal.charon.SkinnableCharonPortal.doGet(SkinnableCharonPortal.java:850)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:622)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:291)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.sakaiproject.util.RequestFilter.doFilter(RequestFilter.java:467)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:219)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:502)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:142)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:617)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:518)
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1091)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:668)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1521)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1478)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.UnsupportedOperationException
	at java.util.Collections$SingletonList.removeIf(Collections.java:4826)
	at org.sakaiproject.component.gradebook.GradebookServiceHibernateImpl.getGradesForStudentsForItem(GradebookServiceHibernateImpl.java:1762)
	at sun.reflect.GeneratedMethodAccessor549.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:281)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:207)
	at com.sun.proxy.$Proxy43.getGradesForStudentsForItem(Unknown Source)
	at org.sakaiproject.gradebookng.business.GradebookNgBusinessService.buildGradeMatrix(GradebookNgBusinessService.java:716)
	at WICKET_org.sakaiproject.gradebookng.business.GradebookNgBusinessService$$FastClassByCGLIB$$86be4ca5.invoke(<generated>)
	at net.sf.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.apache.wicket.proxy.LazyInitProxyFactory$CGLibInterceptor.intercept(LazyInitProxyFactory.java:333)
	at WICKET_org.sakaiproject.gradebookng.business.GradebookNgBusinessService$$EnhancerByCGLIB$$1a5ce107.buildGradeMatrix(<generated>)
	at org.sakaiproject.gradebookng.tool.panels.InstructorGradeSummaryGradesPanel.onInitialize(InstructorGradeSummaryGradesPanel.java:68)
	at org.apache.wicket.Component.fireInitialize(Component.java:877)
	at org.apache.wicket.MarkupContainer.internalInitialize(MarkupContainer.java:961)
	at org.apache.wicket.MarkupContainer.addedComponent(MarkupContainer.java:938)
	at org.apache.wicket.MarkupContainer.replace(MarkupContainer.java:746)
	at org.apache.wicket.MarkupContainer.addOrReplace(MarkupContainer.java:209)
	at org.apache.wicket.extensions.markup.html.tabs.TabbedPanel.setCurrentTab(TabbedPanel.java:431)
	at org.apache.wicket.extensions.markup.html.tabs.TabbedPanel.setSelectedTab(TabbedPanel.java:389)
	at org.apache.wicket.extensions.markup.html.tabs.TabbedPanel.onBeforeRender(TabbedPanel.java:260)
	at org.apache.wicket.Component.internalBeforeRender(Component.java:949)
	at org.apache.wicket.Component.beforeRender(Component.java:1017)
	at org.apache.wicket.MarkupContainer.onBeforeRenderChildren(MarkupContainer.java:1684)
	... 66 more`

Will leave this one open with the "ready to test label until #2732 is resolved.

**Comment by payten** [comment]
Ah yep, thanks @kyleblythe.  I've submitted a PR for #2732, so hopefully that'll do the trick!

**Comment by kyleblythe** [comment]
Looking beautiful - thanks, @payten!

Confirmed and closed.
